tests/cpp/engine/threaded_engine_test.cc

```diff
@@ -254,3 +257,59 @@ TEST(Engine, basics) {
   oprs.clear();
   LOG(INFO) << "All pass";
 }
+
+#ifdef _OPENMP
+
+struct TestSaveAndRestoreOMPState {
+  TestSaveAndRestoreOMPState() {
+    omp_set_dynamic(false);
+  }
+  ~TestSaveAndRestoreOMPState() {
+    omp_set_num_threads(nthreads_);
+    omp_set_dynamic(dynamic_);
+  }
+  const int nthreads_ = omp_get_max_threads();
+  const int dynamic_ = omp_get_dynamic();
+};
+
+/*!
+ * \brief This test checks that omp_set_num_threads implementation has thread-scope
+ */
+TEST(Engine, omp_threading_count_scope) {
+  TestSaveAndRestoreOMPState omp_state;
+  const int THREAD_COUNT = 10;
+  std::shared_ptr<dmlc::ManualEvent> ready = std::make_shared<dmlc::ManualEvent>();
+  std::shared_ptr<dmlc::ThreadGroup> threads = std::make_shared<dmlc::ThreadGroup>();
+  std::atomic<int> counter(0), correct(0);
+  omp_set_dynamic(0);
+  for (int x = 0; x < THREAD_COUNT; ++x) {
+    std::string name = "thread: ";
+    name += std::to_string(x + 1);
+    ++counter;
+    threads->create(name, false,
+                    [x, &counter, &correct](std::shared_ptr<dmlc::ManualEvent> ready_ptr) -> int {
+                      const int thread_count = x + 1;
```

Comment:
Each thread sets omp_num_threads to a different number, then waits for time too start